File: week4/helper(chaltteok)/chaltteok.cpp
#include <cstdio>
#include <vector>
#include <cstdlib>

int pack(std::vector<int> &M) {
    printf("? %d", (int)M.size());
    for(int &x : M) printf(" %d", x);
    puts("");
    fflush(stdout);
    int ret; scanf("%d", &ret);
    if(ret == -202) exit(0);
    return ret;
}

void report(std::vector<int> &M) {
    printf("!");
    for(int &x : M) printf(" %d", x);
    puts("");
    fflush(stdout);
}

int main() {
    int N; scanf("%d", &N);
    std::vector<int> sel, ans;
    
    /* IMPLEMENT HERE */
    /* IMPLEMENT HERE */
    /* IMPLEMENT HERE */
    /* IMPLEMENT HERE */
    // Below is an example.
    // Don't forget to initialize all global variables.

    int turn = 0;
    
    while(turn < 3){
        sel.clear();
        for(int i=1;i<=N;i++){
            sel.push_back(i);
        }
        int ret = pack(sel);
        turn++;
    }
    for(int i=1;i<=N;i++){
        ans.push_back(i);
    }
    report(ans);
    return 0;
}

File: week4/helper(chaltteok)/chaltteok.java
import java.util.Scanner;
import java.util.ArrayList;
import java.util.List;
import java.util.Iterator;

public class chaltteok {
	static Scanner sc;

	static int pack(List<Integer> M) {
		System.out.print("?");
		System.out.print(" " + M.size());
		Iterator it = M.iterator();
		while(it.hasNext()){
			System.out.print(" " + it.next());
		}
		System.out.println();
		System.out.flush();
		int line = sc.nextInt();
		if(line == -202){
			System.exit(0);	
		}
		return line;
	}

	static void report(List<Integer> M) {
		System.out.print("!");
		Iterator it = M.iterator();
		while(it.hasNext()){
			System.out.print(" " + it.next());
		}
		System.out.println();
		System.out.flush();
	}

	public static void main(String[] args) {
		sc = new Scanner(System.in);

		int N = sc.nextInt();

		/* IMPLEMENT HERE */
		/* IMPLEMENT HERE */
		/* IMPLEMENT HERE */
		/* IMPLEMENT HERE */
		// Below is an example.
		// Don't forget to initialize all global variables.

		List<Integer> sel = new ArrayList<Integer>();
		List<Integer> ans = new ArrayList<Integer>();
		
		int turn = 0;
		while(turn < 3){
			sel.clear();
			for(int i = 1; i <= N; i++){
				sel.add(i);
			}
			int ret = pack(sel);
			turn++;
		}
		for(int i = 1; i <= N; i++) {
			ans.add(i);	
		}
		report(ans);
	}
};
